patches/1.15.8.1/LuaJIT-2.1-20190507_03-arm64-lightuserdata.patch

```diff
@@ -0,0 +1,108 @@
+    / 0000000000000000 0111111111111111 1111111111111111 1111111111111111
+    | 0000000000000000 0111111111111111 1111111111111111 1111111111111110
+ 64G| ...
+    | 0000000000000000 0100000000000000 0000000000000000 0000000000000001
+    \ 0000000000000000 0100000000000000 0000000000000000 0000000000000000
+    / 0000000000000000 0011111111111111 1111111111111111 1111111111111111
+    | 0000000000000000 0011111111111111 1111111111111111 1111111111111110
+ 64G| ...
+    | 0000000000000000 0000000000000000 0000000000000000 0000000000000001 ^ heap
+    \ 0000000000000000 0000000000000000 0000000000000000 0000000000000000 |
+
+Notice that the stack grows downwards from the top, and the heap grows upwards
+from the bottom. By losing bit 48, that means that only the bottom 128 GiB
+are addressable, meaning that all pointers to the stack and to static variables
+are invalid, and cause a "bad light userdata pointer" error.
+
+This patch modifies ARM64 support so that instead of losing access to the top
+128 GiB, we lose access of the _middle_ 128 GiB, allowing for heap and stack
+pointers to be valid, up to 64 GiB each. This greatly restores compatibility
+with third-party Lua modules, most of which do not expect over 64 GiB of
+addressable space but _do_ expect to be able to store lightuserdata of static
+variables: it allows, for example, lua-cjson 2.1.0.6 to run unmodified with
+LuaJIT on ARM64 (because it stores lightuserdata pointing to static
+variables).
+
+How this patch works:
+
+LuaJIT does a check and produces a "bad light userdata pointer" if any bit
+from bits 48 to 64 are set.
+
+On input (`lua_pushlightuserdata`), this patch checks bits 47 and 48:
+
+* 00 - do nothing, it will be a LuaJIT-accepted value
+* 01 - convert it to 11, so that LuaJIT rejects the pointer
+* 10 - do nothing, LuaJIT will reject the pointer
+* 11 - convert it to 01, so that LuaJIT accepts the pointer
+
+In other words, if bit 47 is set, flip bit 48.
+
+On output (`lua_touserdata`), this patch checks bit 47, and if it is
+set, it sets bit 48 as well.
+
+Signed-off-by: Hisham Muhammad <hisham@gobolinux.org>
+
+diff -ur a/LuaJIT-2.1-20190507/src/lj_api.c b/LuaJIT-2.1-20190507/src/lj_api.c
+--- a/LuaJIT-2.1-20190507/src/lj_api.c	2019-05-02 17:58:14.000000000 -0300
++++ b/LuaJIT-2.1-20190507/src/lj_api.c	2019-07-17 11:48:35.894843435 -0300
+@@ -594,9 +594,14 @@
+   cTValue *o = index2adr(L, idx);
+   if (tvisudata(o))
+     return uddata(udataV(o));
+-  else if (tvislightud(o))
+-    return lightudV(o);
+-  else
++  else if (tvislightud(o)) {
++    void *p = lightudV(o);
++#if LJ_TARGET_ARM64 
++    if ((uintptr_t) p >> 46)
++      p = (void*) ((uintptr_t) (p) | (1UL << 47));
++#endif
++    return p;
++  } else
+     return NULL;
+ }
+ 
+@@ -696,6 +701,11 @@
+ 
+ LUA_API void lua_pushlightuserdata(lua_State *L, void *p)
+ {
++#if LJ_TARGET_ARM64
++  if ((uintptr_t) p & (1UL << 46)) {
++      p = (void*) ((uintptr_t) p ^ (1UL << 47));
```

Comment:
Style: we should add a space between `void` and `*` (`void *`) to stay consistent with rest of the LuaJIT code, such as https://github.com/Kong/openresty-patches/pull/47/files#diff-3c2314e12ede05c7b300085d7e7803c6R99.